### DISCUSSION TITLE
ESP32 MQTT using MicroPython:  Topic, MSG not publishing to MQTT Server hosted on a Raspberry Pi 3b

### DISCUSSION THREAD

**Discussion by marinehm** [micropython/micropython#11917] [ESP32]
I'm still having problems getting a topic, msg published to my MQTT server.
Here's what I have so far:   Not having any luck with the "Publish"
```

import machine
import network
import time
from time import sleep
import utime
from umqtt.simple import MQTTClient

wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.isconnected()
print("Connecting to router...")
wlan.connect('MySSID', 'MyPassword')
time.sleep(1)
print("Connected to WiFi")
print('network config:', wlan.ifconfig())

sta = network.WLAN(network.STA_IF)
if not sta.isconnected():
    sta.active(True)
    print("WiFi !=connected")

mqtt_server = '192.168.1.13'
client_id = 'ESP32C3_Mini'

def mqtt_connect():
    client = MQTTClient(client_id, mqtt_server, keepalive=3600)
    client.connect()
    print('Connected to MQTT Broker @ %s' %(mqtt_server))
    return client
    client.publish(topic_pub, topic_msg)
    sleep(1)
   
def reconnect():
    print('Failed to connect to MQTT.  Reconnecting...')
    time.sleep(1)
    machine.reset()
   
def publish():
    client_publish(topic, value)

try:
    client = mqtt_connect()
   
    while True:
        print("Hello World")
        #send to MQTT Server
        publish(Relay, 1)
        sleep(1)

except OSError as e:
    reconnect()
```

**Comment by marinehm** [comment]
When I comment out:
while True:
    #publish(Relay, 1)

It connects to my network and MQTT Server

**Comment by karfas** [comment]
Please enclose your code in three backticks (see also #9111).

The code you posted is incomplete. E.g. `Relay` isn't defined, `client_publish` (note the underscore) in `publish()`  isn't defined. 
Just check the error messages you get (maybe comment out all try/except handlers) and fix your code.

**Comment by glenn20** [comment]
As pointed out by @karfas, your code is clearly incomplete. `Relay` is undefined, as is `client_publish()` (did you mean `client.publish()` - but `client` is not defined in that scope.) Also, `publish()` is defined with no parameters, but you call it with 2 arguments, and then within `publish()` you call `client_publish()` with two arguments that have not been defined in that scope.

You should try removing the `try..except` clause around your code. If there is an `OSError` raised for some other reason when you call `publish()` you will ignore the error mssage when it might explain the problem.

**Reply by marinehm** [reply]
I got it working this morning before work.  I ran out of time to provide an update.  As soon as I can(remember to do so), I’ll post my code.  Now to find a reason to post.  I have temp sensors all over the house as it is, those are too easy.  Maybe an all in one weather data and mail notification with light beacon at the top to turn on and off via web server.

**Reply by marinehm** [reply]
Completed:  
```# Connect to the LAN, publish MQTT Topic, Msg
#
# By Nick Sebring;  Jul 4th 2023
# Credit goes to Bhavesh Kakwani: https://bhave.sh/micropython-mqtt/?fbclid=IwAR3Bh5AEMqOEI7QAAtvy_HQTRPfjZeaU9KGbgvLMruiq-WjZ-O-qgaog218
#
#
#  Boot.py connects to Home WiFi
#  Rename this as main.py on microcontroller (ESP32)


from umqtt.simple import MQTTClient
from machine import Pin
from time import sleep

#  Connect to Mqtt broker
CLIENT_NAME = 'ESP32S3Pro'
BROKER_ADDR = '192.168.1.13'
mqttc = MQTTClient(CLIENT_NAME, BROKER_ADDR, keepalive=60)
mqttc.connect()

#  Do Something  #

while True:
    mqttc.publish('Remote/Relay', '1')
    sleep(3600) #3600 = 1 hour```

**Comment by marinehm** [comment]
A year or so ago, I posted the question:  now I don’t remember.  It was either; can you pub or sub more than one topic per client. Back then,  you could not?   Still hold true?    I’m going to give it a try.  I’ll be back……

**Reply by marinehm** [reply]
Yes.  It can publish more than one topic,msg.
Playing around to see if I can Publish and Subscribe (Pub-Sub) all at the same time.  I see it can be done.  I'm going to try it on my own until exhausted with trial and error.

**Reply by marinehm** [reply]
Yes, it does.  Now to clean it up and separate functions.  Read and Publish Temp every 10 minutes and subscribe to Remote/Relay.
```# Connect to the LAN, publish MQTT Topic, Msg
#
# By Nick Sebring;  Jul 4th 2023


#  Boot.py connects to Home WiFi
#  Rename this as main.py

from umqtt.simple import MQTTClient
from machine import Pin
from time import sleep

#  Connect to Mqtt broker
CLIENT_NAME = 'ESP32S3Pro'  # each client must have a unique name
BROKER_ADDR = '192.168.1.13'
mqtt = MQTTClient(CLIENT_NAME, BROKER_ADDR, keepalive=60)
mqtt.connect()

#create object
led = Pin(2, Pin.OUT)

LED_TOPIC= 'Remote/Relay'

def blink_led(topic, msg):
    if msg.decode() == '1':
        led.value(1)
        print(topic,msg)
    else:
        led.value(0)
        print(topic,msg)
#MQTT Subscribe
mqtt.set_callback(blink_led)
mqtt.subscribe(LED_TOPIC)


#  Do Something  #
while True:
    mqtt.publish('Upstairs/Temp', '360')
    mqtt.subscribe('Remote/Relay')
    mqtt.check_msg()
    sleep(.5)```
